src/js/common/containers/DowntimeNotification.jsx

```diff
@@ -0,0 +1,248 @@
+  es: 'es',
+  evss: 'evss',
+  idme: 'idme',
+  mvi: 'mvi',
+  mhv: 'mhv',
+  tims: 'tims',
+  vic: 'vic'
+};
+
+export const serviceStatus = {
+  down: 'down',
+  downtimeApproaching: 'downtimeApproaching',
+  ok: 'ok'
+};
+
+function DowntimeNotificationWrapper({ status, children }) {
+  return <div id="downtime-notification" data-status={status} style={{ marginBottom: '1em' }} className="row-padded">{children}</div>;
+}
+
+class DowntimeNotification extends React.Component {
+
+  static propTypes = {
+    appTitle: PropTypes.string,
+    children: PropTypes.node,
+    content: PropTypes.node,
+    dependencies: PropTypes.arrayOf(PropTypes.oneOf(objectValues(services))).isRequired,
+    determineStatus: PropTypes.func,
+    getScheduledDowntime: PropTypes.func.isRequired,
+    isReady: PropTypes.bool,
+    loadingIndicator: PropTypes.node,
+    render: PropTypes.func,
+    scheduledDowntime: PropTypes.arrayOf(
+      PropTypes.shape({
+        service: PropTypes.string,
+        description: PropTypes.string,
+        startTime: PropTypes.instanceOf(Date),
+        endTime: PropTypes.instanceOf(Date)
+      })
+    ),
+    userIsAuthenticated: PropTypes.bool
+  };
+
+  static defaultProps = {
+    appTitle: 'application'
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      modalDismissed: false,
+      cache: {}
+    };
+  }
+
+  componentWillMount() {
+    this.props.getScheduledDowntime();
+  }
+
+  // This is here just for caching the result of calculateDowntime, so that it isn't run every time a prop changes.
+  // Currently, this component should only do its calculations once, because it would be weird if in the middle of filling
+  // out a form a modal or alert appears alerting them of downtime, or if it just shuts down the app because we entered
+  // scheduled downtime.
+  componentWillReceiveProps(newProps) {
+    const firstLoad = newProps.isReady && !this.props.isReady;
+    if (firstLoad) {
+      const downtimeMap = this.calculateDowntime(newProps.dependencies, newProps.scheduledDowntime);
+      const status = this.determineStatus(downtimeMap);
+      const downtimeWindow = this.getDowntimeWindow(downtimeMap[status]);
+      const cache = { downtimeMap, status, downtimeWindow };
+      this.setState({ cache });
+    }
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    if (!nextProps.isReady) return false;
+    if (nextProps.isReady && !this.props.isReady) return true;
+
+    return nextState.cache.status !== serviceStatus.down;
+  }
+
+  getStatusForDowntime(downtime, current = moment(), warning = moment().add(1, 'hour')) {
+    const inclusive = '[]';
+
+    if (!downtime) return serviceStatus.ok;
+    if (current.isBetween(downtime.startTime, downtime.endTime, inclusive)) return serviceStatus.down;
+    if (warning.isBetween(downtime.startTime, downtime.endTime, inclusive)) return serviceStatus.downtimeApproaching;
+
+    return serviceStatus.ok;
+  }
+
+  getDowntimeWindow(downtimes) {
```

Comment:
So if we have something like
```js
downtimes = [
  { startTime: '10:00', endTime: '11:00' },
  { startTime: '12:00', endTime: '13:00' }
}
```
then `getDowntimeWindow(downtimes)` will return `{ startTime: '11:00', endTime: '13:00' }`, is that right?
(This took me entirely too long to figure out what the purpose was. :grimacing: )

Reply:
Yeah exactly! It has the full downtime object as returned from the action [here](http://jenkins.vetsgov-internal/blue/organizations/jenkins/testing%2Fvets-website/detail/scheduled_downtime/13/pipeline), but it only needs that properties. This may have been overkill, but it's ultimately finding the earliest start time and the latest end time from the application's dependencies and returning that as the downtime window.

Is this too cryptic? I tried to write it clearly but it was one of those things were it doesn't seem like there was a super clear way to do it.

Reply:
Sort of - it's ultimately just trying to find the earliest start time and the latest end time. This is probably overkill but I figured it's also the most correct way to do it in case there's more than one outage. I might rewrite this with variables named more explicitly for `earliestStart`.

Reply:
Do we know what we'd expect to get from the api if we had downtimes on the same day, but like 10 hours apart? Would it give us both? I'm wondering if this means it'll say we've got a downtime of 11 hours, when we really only have two 1-hour downtimes.

Reply:
That is a really great point, nice find. I suppose the pattern it should follow is finding the earliest start point, and the end point based on continuous outage...or maybe that's overkill and it should just be earliest start point and just use the corresponding end point.

Reply:
I vote for the latter - just display the next upcoming outage for a given service. I doubt we'll have windows defined in rapid succession or overlapping, but we might, say, populate multiple distinct windows coming up in the next week when we hear about them.

Reply:
That sounds good to me!